src/domain/user/create-users-with-csv.use-case.ts

```diff
@@ -0,0 +1,113 @@
+import { Command } from '@core/contracts';
+import { PasswordService } from '@core/security';
+import { CsvService } from '@core/security/csv/csv.service';
+import { InputModelValidationService } from '@core/validation/input-model-validation.service';
+import { AddressDbDataSource } from '@data/address/address.db.data-source';
+import { UserDbDataSource } from '@data/user/user.db.data-source';
+import { InvalidFileError, UserAlreadyExistsError } from '@domain/error';
+import { FileModel, UserWithAddressModel, UserCsvModel, UserInputModel, AddressInputModel } from '@domain/model';
+import { Service } from 'typedi';
+
+export type CreateUsersWithCsvUseCaseProps = {
+  fileModel: FileModel;
+  validatorClass: new () => object;
+};
+
+export type EmailWithRawPassword = {
+  email: string;
+  rawPassword: string;
+};
+
+export type UserReadyToSaveWithAddress = {
+  user: UserInputModel;
+  address: AddressInputModel[];
+};
+
+@Service()
+export class CreateUsersWithCsvUseCase implements Command<CreateUsersWithCsvUseCaseProps, UserWithAddressModel[]> {
+  constructor(
+    private readonly userDataSource: UserDbDataSource,
+    private readonly addressDataSource: AddressDbDataSource,
+    private readonly csvService: CsvService,
+    private readonly validationService: InputModelValidationService,
+  ) {}
+
+  async execute(props: CreateUsersWithCsvUseCaseProps): Promise<UserWithAddressModel[]> {
+    const { fileModel: data, validatorClass } = props;
+
+    if (data.mimetype !== 'text/csv') {
+      throw new InvalidFileError('File uploaded with invalid format! Expect csv file.');
+    }
+
+    const entryObjects = await this.csvService.toObject(data.readStream);
+
+    const usersReadyToSave = entryObjects.map(async (rawObj) => {
+      const { hashedPassword } = PasswordService.generateRandomPassword();
+      const userModel = await this.generateAValidUserCsvInstance(rawObj, validatorClass);
+      const userReadyToSave = await this.convertUserCsvToUserReadyToSave(hashedPassword, userModel);
+
+      return userReadyToSave;
+    });
```

Comment:
se for um csv muuito grande, o que pode acontecer com essa array de promises?

Reply:
Se tiver muitos itens no array provavelmente vai sobrecarregar o event loop e dar memory leak. A forma de corrigir isso seria transformando esse array de promises em um for com await, o que processaria cada promise por fez perdendo performance, ou então implementar um batching como:

```typescript
  let results = [];
  for (let index = 0; index < array.length; index += batchSize) {
    const batch = array.slice(index, index + batchSize);
    const batchResults = await Promise.all(batch.map(callback));
    results = results.concat(batchResults);
  }
  return results;
```

dessa forma eu consigo limitar a quantidade de promisses ao mesmo tempo que consigo lançar mais de uma ao mesmo tempo.

Reply:
[399ab89](https://github.com/walefy/taqtile-bot/pull/29/commits/399ab89958c7307aa9cf58b54c9635f4f9819544)

O problema dessa solução é que pode dar timeout na request, mas acho que a única forma de contornar isso seria falar pro usuário que vamos processar e avisar depois por meio de algo como email o status do processamento